### PR TITLE
Limit unsound indexed access type relations

### LINKED ISSUE #27470: undefined extends all types when accessing a generic in a mapped typed
**TypeScript Version**: `3.2.0-dev.20180929` (First noticed in `3.1.1`)

**`strictNullChecks` must be on** [otherwise this is expected behaviour](https://github.com/Microsoft/TypeScript/wiki/Breaking-Changes#intersecting-with-nullundefined-reduces-to-nullundefined-outside-of-strictnullchecks).

It seems that an issue has been introduced when accessing attributes of a mapped type in a conditional type.

**Search Terms:**
conditional type undefined extends
mapped type undefined extends
conditional type in mapped type access

**Code**

```ts
type UndefinedKeys<T extends Record<string, any>> = {
  [K in keyof T]: undefined extends T[K] ? K : never
};

type MyType = {a: string, b: string | undefined}

type Result1 = UndefinedKeys<MyType>;

const a1: Result1['a'] = 'a'; // $ExpectError
      ~~  There should be an error here
const b1: Result1['b'] = 'b';

// manually mapping the type, error is show as expected
type Result2 = {
  a: undefined extends MyType['a'] ? 'a' : never;
  b: undefined extends MyType['b'] ? 'b' : never;
};

const a2: Result2['a'] = 'a'; // $ExpectError
const b2: Result2['b'] = 'b';
```

To simplify this even more line 5 can be changed to
```ts
type MyType = { a: string, b: undefined };
```

**Expected behavior:**

1. The type `Result1` should be `{a: never; b: 'b'}`.
1. Line 9 should have an error: `Type '"a"' is not assignable to type 'never'.`

**Actual behavior:**

1. The type `Result1` is `{a: 'a'; b: 'b'}`.
1. Line 9 has no error.

**Playground Link:** 

| **TypeScript Version** | Playground Link |
|---|---|
| 2.8.1 | [Playground](https://agentcooper.github.io/typescript-play/?ts=2.8.1#code/C4TwDgpgBAqgdgEwgMwJZwgg0hEBnAHgBUoIAPYCRPKAJQgGMB7AJwQL2BfQHMAaKAEM4IAHyioAXigBvAFBQoAbSxR0UANa4myKEQC6ALigBXRCnSZSFKghpEV+qAH4oq4xgBuEFnIC+ANxycqCQUACyIETg0NIygsac3HD8UABGiVy8UAA+puZoGAh+waHQ9HgmADbAAIxSsAWW2LiEkdGQokFyzHCcQrXGFdV1SgDkgmNO0hNjAVAA9AtQACQAomSQDMBrLCysPUx9wOmDdBCVNbXjaVMNY7fdS1AAtsImglVVIK+CYGDZYAAC2gZRCMXOl2AACYGvJFAl8khClZyJRqBEojFxpMnK5ZlAPBBvCwgooMkiLEVrOi7JiOhAbnd8bdCVAvD4goFgr1+oJoUMLiNoTi7jNJvNnutNowdnsDryTmkBZDhUzplAHnM5EA) |
| 3.0.3 | [Playground](https://agentcooper.github.io/typescript-play/#code/C4TwDgpgBAqgdgEwgMwJZwgg0hEBnAHgBUoIAPYCRPKAJQgGMB7AJwQL2BfQHMAaKAEM4IAHyioAXigBvAFBQoAbSxR0UANa4myKEQC6ALigBXRCnSZSFKghpEV+qAH4oq4xgBuEFnIC+ANxycqCQUACyIETg0NIygsac3HD8UABGiVy8UAA+puZoGAh+waHQ9HgmADbAAIxSsAWW2LiEkdGQokFyzHCcQrXGFdV1SgDkgmNO0hNjAVAA9AtQACQAomSQDMBrLCysPUx9wOmDdBCVNbXjaVMNY7fdS1AAtsImglVVIK+CYGDZYAAC2gZRCMXOl2AACYGvJFAl8khClZyJRqBEojFxpMnK5ZlAPBBvCwgooMkiLEVrOi7JiOhAbnd8bdCVAvD4goFgr1+oJoUMLiNoTi7jNJvNnutNowdnsDryTmkBZDhUzplAHnM5EA) |
| 3.1.1 | [Playground](http://www.typescriptlang.org/play/#src=type%20UndefinedKeys%3CT%20extends%20Record%3Cstring%2C%20any%3E%3E%20%3D%20%7B%0D%0A%20%20%5BK%20in%20keyof%20T%5D%3A%20undefined%20extends%20T%5BK%5D%20%3F%20K%20%3A%20never%0D%0A%7D%3B%0D%0A%0D%0Atype%20MyType%20%3D%20%7Ba%3A%20string%2C%20b%3A%20string%20%7C%20undefined%7D%0D%0A%0D%0Atype%20Result1%20%3D%20UndefinedKeys%3CMyType%3E%3B%0D%0A%0D%0Aconst%20a1%3A%20Result1%5B'a'%5D%20%3D%20'a'%3B%20%2F%2F%20%24ExpectError%0D%0Aconst%20b1%3A%20Result1%5B'b'%5D%20%3D%20'b'%3B%0D%0A%0D%0A%2F%2F%20manually%20mapping%20the%20type%0D%0Atype%20Result2%20%3D%20%7B%0D%0A%20%20a%3A%20undefined%20extends%20MyType%5B'a'%5D%20%3F%20'a'%20%3A%20never%3B%0D%0A%20%20b%3A%20undefined%20extends%20MyType%5B'b'%5D%20%3F%20'b'%20%3A%20never%3B%0D%0A%7D%3B%0D%0A%0D%0Aconst%20a2%3A%20Result2%5B'a'%5D%20%3D%20'a'%3B%20%2F%2F%20%24ExpectError%0D%0Aconst%20b2%3A%20Result2%5B'b'%5D%20%3D%20'b'%3B%0D%0A) <---- ENABLE `strictNullChecks` IN OPTIONS |
| 3.2.0-dev.20180929 | Same error as 3.1.1 |

**Related Issues:**
Potentially: https://github.com/Microsoft/TypeScript/issues/26942

Initially thought it was an issue with [Distributive conditional types](https://www.typescriptlang.org/docs/handbook/advanced-types.html#distributive-conditional-types) but the Generic type is in the assignable to position.

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #27470.

### CODE REVIEW

File: tests/baselines/reference/keyofAndIndexedAccess.js [hunk 1]
```diff
@@ -300,23 +300,16 @@ type S2 = {
     b: string;
 };
 
-function f90<T extends S2, K extends keyof S2>(x1: S2[keyof S2], x2: T[keyof S2], x3: S2[K], x4: T[K]) {
+function f90<T extends S2, K extends keyof S2>(x1: S2[keyof S2], x2: T[keyof S2], x3: S2[K]) {
     x1 = x2;
     x1 = x3;
-    x1 = x4;
```
Comment (weswigham) [inline tests/baselines/reference/keyofAndIndexedAccess.js]: Might be worth keeping all these variants (maybe in a different file or function) just to document that they're expected to fail now (and so we know if they suddenly start succeeding again).

**Comment by ahejlsberg** [comment]
@weswigham Do we have an automated way of running the DT tests? (Similar to the RWC bot.) Otherwise, what's the best way to go about it?

**Comment by weswigham** [comment]
@ahejlsberg No, the bot doesn't have a build for baselining dt. I can  probably whip one up, though I don't think it'll be able to have an automatic PR/diff viewing system like RWC does. In the absence of that, running `jake runtests-parallel ru=dt` on master, accepting the initial baseline snapshot, then running it again on this branch would be a way to go about it locally; but you probably want someone with a real beefy machine to do it, since they take quite a long time compared to RWC, even.

**Comment by ahejlsberg** [comment]
> but you probably want someone with a real beefy machine to do it

Like yours? (Hint, hint)

**Comment by weswigham** [comment]
I knew what I was getting into when I said what I said. XD

**Comment by weswigham** [comment]
And 4 hours later the results are in. My fears seem overblown, this change doesn't seem to affect anything on DT.

@typescript-bot test this just in case tho

**Comment by squirly** [comment]
I just tried this in my project, and it indeed does fix #27470! :tada:

**Comment by weswigham** [comment]
I think rwc on this PR is fine; this PR is just missing a change in master (semicolon in a helper) which caused an RWC change. Everything looks good.

**Pull request merged**